docs/linux-system/06-configuration-and-security.md
# Linux系统配置与安全

在完成文件编辑的学习之后，也就可以重新回到服务相关的内容，去通过了解配置，更改配置文件等方式去使我们常用的服务更快捷，更方便

## SSH

### 配置文件与配置

配置文件的位置在`/etc/ssh/sshd_config`，简单看看配置文件的内容

```ini
# This is the sshd server system-wide configuration file.  See
# sshd_config(5) for more information.

# This sshd was compiled with PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr                                                                  /bin:/sbin:/bin:/usr/games

# The strategy used for options in the default sshd_config shipped with
# OpenSSH is to specify options with their default value where
# possible, but leave them commented.  Uncommented options override the
# default value.

Include /etc/ssh/sshd_config.d/*.conf

# When systemd socket activation is used (the default), the socket
# configuration must be re-generated after changing Port, AddressFamily, or
# ListenAddress.
#
# For changes to take effect, run:
#
#   systemctl daemon-reload
#   systemctl restart ssh.socket
#
#Port 22
#AddressFamily any
#ListenAddress 0.0.0.0
#ListenAddress ::

#HostKey /etc/ssh/ssh_host_rsa_key
#HostKey /etc/ssh/ssh_host_ecdsa_key
#HostKey /etc/ssh/ssh_host_ed25519_key

# Ciphers and keying
#RekeyLimit default none

# Logging
#SyslogFacility AUTH
#LogLevel INFO

# Authentication:

#LoginGraceTime 2m
#PermitRootLogin prohibit-password
#StrictModes yes
#MaxAuthTries 6
#MaxSessions 10

#PubkeyAuthentication yes

# Expect .ssh/authorized_keys2 to be disregarded by default in future.
#AuthorizedKeysFile     .ssh/authorized_keys .ssh/authorized_keys2

#AuthorizedPrincipalsFile none

#AuthorizedKeysCommand none
#AuthorizedKeysCommandUser nobody

# For this to work you will also need host keys in /etc/ssh/ssh_known_hosts
#HostbasedAuthentication no
# Change to yes if you don't trust ~/.ssh/known_hosts for
# HostbasedAuthentication
#IgnoreUserKnownHosts no
# Don't read the user's ~/.rhosts and ~/.shosts files
#IgnoreRhosts yes

# To disable tunneled clear text passwords, change to no here!
#PasswordAuthentication yes
#PermitEmptyPasswords no

# Change to yes to enable challenge-response passwords (beware issues with
# some PAM modules and threads)
KbdInteractiveAuthentication no

# Kerberos options
#KerberosAuthentication no
#KerberosOrLocalPasswd yes
#KerberosTicketCleanup yes
#KerberosGetAFSToken no

# GSSAPI options
#GSSAPIAuthentication no
#GSSAPICleanupCredentials yes
#GSSAPIStrictAcceptorCheck yes
#GSSAPIKeyExchange no

# Set this to 'yes' to enable PAM authentication, account processing,
# and session processing. If this is enabled, PAM authentication will
# be allowed through the KbdInteractiveAuthentication and
# PasswordAuthentication.  Depending on your PAM configuration,
# PAM authentication via KbdInteractiveAuthentication may bypass
# the setting of "PermitRootLogin prohibit-password".
# If you just want the PAM account and session checks to run without
# PAM authentication, then enable this but set PasswordAuthentication
# and KbdInteractiveAuthentication to 'no'.
UsePAM yes

#AllowAgentForwarding yes
#AllowTcpForwarding yes
#GatewayPorts no
X11Forwarding yes
#X11DisplayOffset 10
#X11UseLocalhost yes
#PermitTTY yes
PrintMotd no
#PrintLastLog yes
#TCPKeepAlive yes
#PermitUserEnvironment no
#Compression delayed
#ClientAliveInterval 0
#ClientAliveCountMax 3
#UseDNS no
#PidFile /run/sshd.pid
#MaxStartups 10:30:100
#PermitTunnel no
#ChrootDirectory none
#VersionAddendum none

# no default banner path
#Banner none

# Allow client to pass locale environment variables
AcceptEnv LANG LC_*

# override default of no subsystems
Subsystem       sftp    /usr/lib/openssh/sftp-server

# Example of overriding settings on a per-user basis
#Match User anoncvs
#       X11Forwarding no
#       AllowTcpForwarding no
#       PermitTTY no
#       ForceCommand cvs server
```

---
### 公钥和私钥

**生成密钥**

认证方式上面好像除了现在简单的用户名密码认证还看到了一个被注释掉的选项`PubkeyAuthentication`，这种登录方式相比于账户密码的方式会更快捷和安全，所以还是比较建议打开的

首先需要在本地生成本机的公钥和私钥，打开Windows的PowerShell，在里面输入命令生成密钥与公钥

```powershell
ssh-keygen
```

会显示下面这样的输出

```powershell
Generating public/private ed25519 key pair.
Enter file in which to save the key (C:\Users\Ivo/.ssh/id_ed25519):
```

意思是输入存放公钥与私钥的目录与文件名，默认路径是`C:\Users\Ivo\.ssh\id_ed25519`，如果自定义也是在自己这个用户目录下面

输入后回车进入目录会出现下面两个文件

> - **id_ed25519**
> - **id_ed25519.pub**

其中`.pub`文件是公钥，另一个则是私钥

---

**使用密钥**

本地连接使用私钥，在服务器上存放公钥，进配置文件更改以下行，取消注释即可

```ini
PubkeyAuthentication yes

# Expect .ssh/authorized_keys2 to be disregarded by default in future.
AuthorizedKeysFile      .ssh/authorized_keys .ssh/authorized_keys2
```

```ini
# This is the sshd server system-wide configuration file.  See
# sshd_config(5) for more information.

# This sshd was compiled with PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr                                                                  /bin:/sbin:/bin:/usr/games

# The strategy used for options in the default sshd_config shipped with
# OpenSSH is to specify options with their default value where
# possible, but leave them commented.  Uncommented options override the
# default value.

Include /etc/ssh/sshd_config.d/*.conf

# When systemd socket activation is used (the default), the socket
# configuration must be re-generated after changing Port, AddressFamily, or
# ListenAddress.
#
# For changes to take effect, run:
#
#   systemctl daemon-reload
#   systemctl restart ssh.socket
#
#Port 22
#AddressFamily any
#ListenAddress 0.0.0.0
#ListenAddress ::

#HostKey /etc/ssh/ssh_host_rsa_key
#HostKey /etc/ssh/ssh_host_ecdsa_key
#HostKey /etc/ssh/ssh_host_ed25519_key

# Ciphers and keying
#RekeyLimit default none

# Logging
#SyslogFacility AUTH
#LogLevel INFO

# Authentication:

#LoginGraceTime 2m
#PermitRootLogin prohibit-password
#StrictModes yes
#MaxAuthTries 6
#MaxSessions 10

#PubkeyAuthentication yes // [!code --]
PubkeyAuthentication yes # [!code ++]

# Expect .ssh/authorized_keys2 to be disregarded by default in future.
#AuthorizedKeysFile     .ssh/authorized_keys .ssh/authorized_keys2 // [!code --]
AuthorizedKeysFile     .ssh/authorized_keys .ssh/authorized_keys2 # [!code ++]

#AuthorizedPrincipalsFile none

#AuthorizedKeysCommand none
#AuthorizedKeysCommandUser nobody

# For this to work you will also need host keys in /etc/ssh/ssh_known_hosts
#HostbasedAuthentication no
# Change to yes if you don't trust ~/.ssh/known_hosts for
# HostbasedAuthentication
#IgnoreUserKnownHosts no
# Don't read the user's ~/.rhosts and ~/.shosts files
#IgnoreRhosts yes

# To disable tunneled clear text passwords, change to no here!
#PasswordAuthentication yes
#PermitEmptyPasswords no

# Change to yes to enable challenge-response passwords (beware issues with
# some PAM modules and threads)
KbdInteractiveAuthentication no

# Kerberos options
#KerberosAuthentication no
#KerberosOrLocalPasswd yes
#KerberosTicketCleanup yes
#KerberosGetAFSToken no

# GSSAPI options
#GSSAPIAuthentication no
#GSSAPICleanupCredentials yes
#GSSAPIStrictAcceptorCheck yes
#GSSAPIKeyExchange no

# Set this to 'yes' to enable PAM authentication, account processing,
# and session processing. If this is enabled, PAM authentication will
# be allowed through the KbdInteractiveAuthentication and
# PasswordAuthentication.  Depending on your PAM configuration,
# PAM authentication via KbdInteractiveAuthentication may bypass
# the setting of "PermitRootLogin prohibit-password".
# If you just want the PAM account and session checks to run without
# PAM authentication, then enable this but set PasswordAuthentication
# and KbdInteractiveAuthentication to 'no'.
UsePAM yes

#AllowAgentForwarding yes
#AllowTcpForwarding yes
#GatewayPorts no
X11Forwarding yes
#X11DisplayOffset 10
#X11UseLocalhost yes
#PermitTTY yes
PrintMotd no
#PrintLastLog yes
#TCPKeepAlive yes
#PermitUserEnvironment no
#Compression delayed
#ClientAliveInterval 0
#ClientAliveCountMax 3
#UseDNS no
#PidFile /run/sshd.pid
#MaxStartups 10:30:100
#PermitTunnel no
#ChrootDirectory none
#VersionAddendum none

# no default banner path
#Banner none

# Allow client to pass locale environment variables
AcceptEnv LANG LC_*

# override default of no subsystems
Subsystem       sftp    /usr/lib/openssh/sftp-server

# Example of overriding settings on a per-user basis
#Match User anoncvs
#       X11Forwarding no
#       AllowTcpForwarding no
#       PermitTTY no
#       ForceCommand cvs server
```

`AuthorizedKeysFile`是认证密钥的文件，文件位置在`$HOME/.ssh/authorized_keys`，我们需要将我们`.pub`文件内内容加入进去

```bash
vim $HOME/.ssh/authorized_keys
```

将自己生成的公钥加入进去，多台机器则需要输入多台机器的，例如这样

```ini
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAICZEYTc7XUGMLZX4gEXr6xxHt63xujviUdym6evKzqJy ivo@DESKTOP-BGG1MOL
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIIgwov09dnog5X+JMj+8uSrQBu5CXDpQwvKyxfY/bJa8 ivo@DESKTOP-KLLCI8I
```

重启ssh服务

```bash
systemctl restart sshd
```

在MobaXterm加入自己的私钥

![](assets/06-configuration-and-security.png)

**尝试连接，输入用户名后即可直接连接上自己的机器**

---
### 端口

我们可以回忆一下我们连接自己虚拟机或者实体机的步骤

> - **查看IP**：在Linux系统机器上输入`ip addr`查看机器的IP地址
> - **使用MobaXterm连接**：在MobaXterm中输入机器的IP地址进行连接

但是在连接的时候有一个地方我们完全没有注意——`Port`端口

这个地方默认使用的`22`，这个`22`是默认端口，但是像这样的默认端口在`rdp`，`vnc`，`http`和`https`等协议中都是存在的，这样的端口都有自己的名字——**标准端口**，所以其他的端口可以叫作**非标准端口**，简称**非标端口**

---

但是使用这样的标准端口很容易被发现，所以我们可以更换为`1024-65535`中任何一个没有被占用的端口，只需要修改配置文件里面的`Port`即可

```ini
# This is the sshd server system-wide configuration file.  See
# sshd_config(5) for more information.

# This sshd was compiled with PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr                                                                  /bin:/sbin:/bin:/usr/games

# The strategy used for options in the default sshd_config shipped with
# OpenSSH is to specify options with their default value where
# possible, but leave them commented.  Uncommented options override the
# default value.

Include /etc/ssh/sshd_config.d/*.conf

# When systemd socket activation is used (the default), the socket
# configuration must be re-generated after changing Port, AddressFamily, or
# ListenAddress.
#
# For changes to take effect, run:
#
#   systemctl daemon-reload
#   systemctl restart ssh.socket
#
#Port 22 // [!code --]
Port 2222 # [!code ++]
#AddressFamily any
#ListenAddress 0.0.0.0
#ListenAddress ::

#HostKey /etc/ssh/ssh_host_rsa_key
#HostKey /etc/ssh/ssh_host_ecdsa_key
#HostKey /etc/ssh/ssh_host_ed25519_key

# Ciphers and keying
#RekeyLimit default none

# Logging
#SyslogFacility AUTH
#LogLevel INFO

# Authentication:

#LoginGraceTime 2m
#PermitRootLogin prohibit-password
#StrictModes yes
#MaxAuthTries 6
#MaxSessions 10

PubkeyAuthentication yes

# Expect .ssh/authorized_keys2 to be disregarded by default in future.
AuthorizedKeysFile     .ssh/authorized_keys .ssh/authorized_keys2

#AuthorizedPrincipalsFile none

#AuthorizedKeysCommand none
#AuthorizedKeysCommandUser nobody

# For this to work you will also need host keys in /etc/ssh/ssh_known_hosts
#HostbasedAuthentication no
# Change to yes if you don't trust ~/.ssh/known_hosts for
# HostbasedAuthentication
#IgnoreUserKnownHosts no
# Don't read the user's ~/.rhosts and ~/.shosts files
#IgnoreRhosts yes

# To disable tunneled clear text passwords, change to no here!
#PasswordAuthentication yes
#PermitEmptyPasswords no

# Change to yes to enable challenge-response passwords (beware issues with
# some PAM modules and threads)
KbdInteractiveAuthentication no

# Kerberos options
#KerberosAuthentication no
#KerberosOrLocalPasswd yes
#KerberosTicketCleanup yes
#KerberosGetAFSToken no

# GSSAPI options
#GSSAPIAuthentication no
#GSSAPICleanupCredentials yes
#GSSAPIStrictAcceptorCheck yes
#GSSAPIKeyExchange no

# Set this to 'yes' to enable PAM authentication, account processing,
# and session processing. If this is enabled, PAM authentication will
# be allowed through the KbdInteractiveAuthentication and
# PasswordAuthentication.  Depending on your PAM configuration,
# PAM authentication via KbdInteractiveAuthentication may bypass
# the setting of "PermitRootLogin prohibit-password".
# If you just want the PAM account and session checks to run without
# PAM authentication, then enable this but set PasswordAuthentication
# and KbdInteractiveAuthentication to 'no'.
UsePAM yes

#AllowAgentForwarding yes
#AllowTcpForwarding yes
#GatewayPorts no
X11Forwarding yes
#X11DisplayOffset 10
#X11UseLocalhost yes
#PermitTTY yes
PrintMotd no
#PrintLastLog yes
#TCPKeepAlive yes
#PermitUserEnvironment no
#Compression delayed
#ClientAliveInterval 0
#ClientAliveCountMax 3
#UseDNS no
#PidFile /run/sshd.pid
#MaxStartups 10:30:100
#PermitTunnel no
#ChrootDirectory none
#VersionAddendum none

# no default banner path
#Banner none

# Allow client to pass locale environment variables
AcceptEnv LANG LC_*

# override default of no subsystems
Subsystem       sftp    /usr/lib/openssh/sftp-server

# Example of overriding settings on a per-user basis
#Match User anoncvs
#       X11Forwarding no
#       AllowTcpForwarding no
#       PermitTTY no
#       ForceCommand cvs server
```


## 用户权限

在用户类型和用户权限章节中，`sudo`可以临时给普通用户提权至`root`，所以也就出现了一个问题——用户提权后的操作是否会影响系统的稳定性，所以我们要对用户提权后的操作进行相关的限制

这个是当时的源文件

```ini
#
# This file MUST be edited with the 'visudo' command as root.
#
# Please consider adding local content in /etc/sudoers.d/ instead of
# directly modifying this file.
#
# See the man page for details on how to write a sudoers file.
#
Defaults        env_reset
Defaults        mail_badpass
Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin"

# This fixes CVE-2005-4890 and possibly breaks some versions of kdesu
# (#1011624, https://bugs.kde.org/show_bug.cgi?id=452532)
Defaults        use_pty

# This preserves proxy settings from user environments of root
# equivalent users (group sudo)
#Defaults:%sudo env_keep += "http_proxy https_proxy ftp_proxy all_proxy no_proxy"

# This allows running arbitrary commands, but so does ALL, and it means
# different sudoers have their choice of editor respected.
#Defaults:%sudo env_keep += "EDITOR"

# Completely harmless preservation of a user preference.
#Defaults:%sudo env_keep += "GREP_COLOR"

# While you shouldn't normally run git as root, you need to with etckeeper
#Defaults:%sudo env_keep += "GIT_AUTHOR_* GIT_COMMITTER_*"

# Per-user preferences; root won't have sensible values for them.
#Defaults:%sudo env_keep += "EMAIL DEBEMAIL DEBFULLNAME"

# "sudo scp" or "sudo rsync" should be able to use your SSH agent.
#Defaults:%sudo env_keep += "SSH_AGENT_PID SSH_AUTH_SOCK"

# Ditto for GPG agent
#Defaults:%sudo env_keep += "GPG_AGENT_INFO"

# Host alias specification

# User alias specification

# Cmnd alias specification

# User privilege specification
root    ALL=(ALL:ALL) ALL

# Members of the admin group may gain root privileges
%admin ALL=(ALL) ALL

# Allow members of group sudo to execute any command
%sudo   ALL=(ALL:ALL) ALL

# See sudoers(5) for more information on "@include" directives:

@includedir /etc/sudoers.d
```

---
### 所有的用户改为只能关机重启

```ini
#
# This file MUST be edited with the 'visudo' command as root.
#
# Please consider adding local content in /etc/sudoers.d/ instead of
# directly modifying this file.
#
# See the man page for details on how to write a sudoers file.
#
Defaults        env_reset
Defaults        mail_badpass
Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin"

# This fixes CVE-2005-4890 and possibly breaks some versions of kdesu
# (#1011624, https://bugs.kde.org/show_bug.cgi?id=452532)
Defaults        use_pty

# This preserves proxy settings from user environments of root
# equivalent users (group sudo)
#Defaults:%sudo env_keep += "http_proxy https_proxy ftp_proxy all_proxy no_proxy"

# This allows running arbitrary commands, but so does ALL, and it means
# different sudoers have their choice of editor respected.
#Defaults:%sudo env_keep += "EDITOR"

# Completely harmless preservation of a user preference.
#Defaults:%sudo env_keep += "GREP_COLOR"

# While you shouldn't normally run git as root, you need to with etckeeper
#Defaults:%sudo env_keep += "GIT_AUTHOR_* GIT_COMMITTER_*"

# Per-user preferences; root won't have sensible values for them.
#Defaults:%sudo env_keep += "EMAIL DEBEMAIL DEBFULLNAME"

# "sudo scp" or "sudo rsync" should be able to use your SSH agent.
#Defaults:%sudo env_keep += "SSH_AGENT_PID SSH_AUTH_SOCK"

# Ditto for GPG agent
#Defaults:%sudo env_keep += "GPG_AGENT_INFO"

# Host alias specification

# User alias specification

# Cmnd alias specification

# User privilege specification
root    ALL=(ALL:ALL) ALL

# Members of the admin group may gain root privileges
%admin ALL=(ALL) ALL

# Allow members of group sudo to execute any command
%sudo   ALL=(ALL:ALL) ALL # [!code --]
%sudo   ALL=(ALL:ALL) /sbin/reboot, /sbin/shutdown # [!code ++]

# See sudoers(5) for more information on "@include" directives:

@includedir /etc/sudoers.d
```

---
### 某一个用户获得所有权

```ini
#
# This file MUST be edited with the 'visudo' command as root.
#
# Please consider adding local content in /etc/sudoers.d/ instead of
# directly modifying this file.
#
# See the man page for details on how to write a sudoers file.
#
Defaults        env_reset
Defaults        mail_badpass
Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin"

# This fixes CVE-2005-4890 and possibly breaks some versions of kdesu
# (#1011624, https://bugs.kde.org/show_bug.cgi?id=452532)
Defaults        use_pty

# This preserves proxy settings from user environments of root
# equivalent users (group sudo)
#Defaults:%sudo env_keep += "http_proxy https_proxy ftp_proxy all_proxy no_proxy"

# This allows running arbitrary commands, but so does ALL, and it means
# different sudoers have their choice of editor respected.
#Defaults:%sudo env_keep += "EDITOR"

# Completely harmless preservation of a user preference.
#Defaults:%sudo env_keep += "GREP_COLOR"

# While you shouldn't normally run git as root, you need to with etckeeper
#Defaults:%sudo env_keep += "GIT_AUTHOR_* GIT_COMMITTER_*"

# Per-user preferences; root won't have sensible values for them.
#Defaults:%sudo env_keep += "EMAIL DEBEMAIL DEBFULLNAME"

# "sudo scp" or "sudo rsync" should be able to use your SSH agent.
#Defaults:%sudo env_keep += "SSH_AGENT_PID SSH_AUTH_SOCK"

# Ditto for GPG agent
#Defaults:%sudo env_keep += "GPG_AGENT_INFO"

# Host alias specification

# User alias specification

# Cmnd alias specification

# User privilege specification
root    ALL=(ALL:ALL) ALL
teach   ALL=(ALL:ALL) ALL # [!code ++]

# Members of the admin group may gain root privileges
%admin ALL=(ALL) ALL

# Allow members of group sudo to execute any command
%sudo   ALL=(ALL:ALL) ALL
%sudo   ALL=(ALL:ALL) /sbin/reboot, /sbin/shutdown

# See sudoers(5) for more information on "@include" directives:

@includedir /etc/sudoers.d
```

## 编辑器

Vim相较于Nano会复杂很多，而且不同人使用键盘的习惯是不一样的，快捷键也不是所有人都会觉得顺手，所以更换快捷键也是一个很重要的内容

---
### Vim的配置文件

Vim的配置有两种，一个是全局配置，另一个是用户配置，全局的只能存在一个，每个用户都可以有自己的配置

> - **全局**：`/etc/vim/vimrc`
> - **用户**：`$HOME/.vimrc`

---

### 配置文件的编辑

设置自定义按键：

```ini
# F3 执行:Explore<CR>

map <F3> :Explore<CR>
```

更换特定模式已定义的快捷键

```ini
# 使用jk代替Esc

imap jk <Esc>
```

高亮显示

```ini
syntax on
```

**我的配置文件：**

```ini
"=========================================================================
"
"" 根据他人配置进行增加个性化设置
"
""=========================================================================

map <F3> :Explore<CR>
map <S-F3> :Vexplore<CR>
map <C-F3> :Sexplore<CR>  

imap jk <Esc>

nmap <space> :

syntax on
```


docs/linux-system/07-network-configuration-and-management.md
# 网络配置与管理

在Ubuntu Server的安装中，在安装界面就已经提示要求配置本机的IP地址，当时是直接选择的静态IP，这样我们在开机的时候非常顺利的使用SSH连接上，但是网络配置上Ubuntu的网络框架较为新，所以传统框架需要使用Debian，下载安装最新版本的Debian 13安装镜像即可

## Debian 13安装

我们需要重新安装一个全新的Debian 13虚拟机，网络安装镜像或者本地DVD镜像都可以

Debian安装时的配置比Ubuntu Server少很多，只有一个方面需要注意，也就是软件选择上，按照下图进行选择即可

![](assets/07-network-configuration-and-management.png)

## 传统网络框架 —— `ifupdown`

这个框架的配置文件格式非常清晰，但是因为Ubuntu Server上使用Network Manager管理框架，所以我们安装默认使用这个框架的Debian 13，Debian 13默认创建root账户，我们这里直接使用root账号登录

### 配置文件位置

`ifupdown`框架的配置文件位置在`/etc/network/interfaces`中，Debian 13默认安装`nano`文本编辑器，所以我们先使用文本编辑器打开这个配置文件

```bash
nano /etc/network/interfaces
```

---
### 配置文件查看与理解

配置文件内容如下：

```ini
# This file describes the network interfaces available on your system
# and how to activate them. For more information, see interfaces(5).

source /etc/network/interfaces.d/*

# The loopback network interface
auto lo
iface lo inet loopback

# The primary network interface
allow-hotplug ens18
iface ens18 inet dhcp
```

里面一共写了两张网卡的配置，`lo`是本地回环网卡，也就是常见的`127.0.0.1`，为了确保剩下一张是我们联网的网卡，可以输入这个命令查看

```bash
ip addr
```

系统返回如下

```bash
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute
       valid_lft forever preferred_lft forever
2: ens18: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether bc:24:11:c1:b9:7b brd ff:ff:ff:ff:ff:ff
    altname enp0s18
    altname enxbc2411c1b97b
    inet 192.168.1.108/24 brd 192.168.1.255 scope global dynamic noprefixroute ens18
       valid_lft 85642sec preferred_lft 74842sec
    inet6 fd3e:7e16:83e6::225/128 scope global dynamic noprefixroute
       valid_lft 42439sec preferred_lft 42439sec
    inet6 fd3e:7e16:83e6:0:2989:c87d:c10f:ae89/64 scope global dynamic mngtmpaddr noprefixroute
       valid_lft 4813sec preferred_lft 2113sec
    inet6 fe80::79f5:5e4:4363:5ebc/64 scope link
       valid_lft forever preferred_lft forever
```

说明现在我们总共两张网卡，一个本地回环的`lo`，另一个则是正常负责上网的`ens18`网卡，所以我们需要设置`ens18`这张网卡的静态IP地址

---

### 配置文件的编辑

我们需要对配置文件做出相对应的修改，首先将原本的动态DHCP换成静态的固定IP地址，以及配置对应的IP地址，网管，DNS等

```ini
# This file describes the network interfaces available on your system
# and how to activate them. For more information, see interfaces(5).

source /etc/network/interfaces.d/*

# The loopback network interface
auto lo
iface lo inet loopback

# The primary network interface
allow-hotplug ens18
iface ens18 inet dhcp # [!code --]
iface ens18 inet static # [!code ++]
	address 192.168.1.10/24 # [!code ++]
	gateway 192.168.1.1 # [!code ++]
	dns-nameserver 223.5.5.5 # [!code ++]
```

但是因为`allow-hotplug`在配置文件中，这个网卡是只有在重新插拔之后配置才能生效，所以为了能够最快时间生效，我们需要将这个改为`auto`

```ini
# This file describes the network interfaces available on your system
# and how to activate them. For more information, see interfaces(5).

source /etc/network/interfaces.d/*

# The loopback network interface
auto lo
iface lo inet loopback

# The primary network interface
allow-hotplug ens18 # [!code --]
auto ens18 # [!code ++]
iface ens18 inet static
	address 192.168.1.10/24
	gateway 192.168.1.1
	dns-nameserver 223.5.5.5
```

---

### 使配置文件生效

重启网络服务，使网络配置重新加载

```bash
sudo systemctl restart networking
```

---
### 检查IP生效情况

```bash
ip addr
```

## 现代网络框架 —— `netplan`

相较于传统的网络配置，`netplan`最明显的就是他的配置文件格式，从以前的`ini`格式换成了更简洁而且容易看懂的`yaml`格式

---

### 配置文件的位置

配置文件通常位于`/etc/netplan/`这个目录下，不同的机器安装之后这个文件名称也不一样，但是通过查看目录下配置文件就可以看到

```bash
ls /etc/netplan/
```

系统会显示一个`yaml`文件

```bash
50-cloud-init.yaml
```

---

### 配置文件查看与理解

在编辑这个文件之前，我们需要看一下这个文件内的具体内容

```bash
sudo cat /etc/netplan/50-cloud-init.yaml
```

返回文件内的具体内容

```yaml
network:
  version: 2
  ethernets:
    ens18:
      addresses:
      - "192.168.1.4/24"
      nameservers:
        addresses:
        - 223.5.5.5
        search: []
      routes:
      - to: "default"
        via: "192.168.1.1"
```

里面能很清晰看到我们一张网卡的信息

> - **ens18**：网卡的名称
> - **addresses**：IP地址，但是需要写成带子网掩码的格式例如子网掩码是`255.255.255.0`要写成`/24`
> - **namservers**：DNS服务器
> - **routes**：路由，也就是网关`gateway`

---

### 编辑配置文件

如果我想将IP地址从`192.168.1.4`换成`192.168.1.10`

```yaml
network:
  version: 2
  ethernets:
    ens18:
      addresses:
      - "192.168.1.4/24" # [!code --]
      - "192.168.1.10/24" # [!code ++]
      nameservers:
        addresses:
        - 223.5.5.5
        search: []
      routes:
      - to: "default"
        via: "192.168.1.1"
```

---

### 使配置文件生效

`netplan`的生效方式不一样，`netplan`本质上是一个一次性命令行工具，不需要系统的守护进程，所以我们需要使`netplan`生效即可

```bash
sudo netplan apply
```

---

### 检查IP生效情况

```bash
ip addr
```

.vitepress/config.mjs
import { defineConfig } from "vitepress";

export default defineConfig({
    srcDir: "docs",
    publicDir: ".vitepress/public",
    head: [
        ["link", { rel: "icon", type: "image/svg+xml", href: "/favicon.svg" }],
        [
            "link",
            {
                rel: "icon",
                type: "image/png",
                sizes: "16x16",
                href: "/favicon-16x16.png",
            },
        ],
        [
            "link",
            {
                rel: "icon",
                type: "image/png",
                sizes: "32x32",
                href: "/favicon-32x32.png",
            },
        ],
        [
            "link",
            {
                rel: "icon",
                type: "image/png",
                sizes: "48x48",
                href: "/favicon-48x48.png",
            },
        ],
        [
            "link",
            {
                rel: "apple-touch-icon",
                sizes: "180x180",
                href: "/apple-touch-icon.png",
            },
        ],
        [
            "link",
            {
                rel: "apple-touch-icon",
                sizes: "152x152",
                href: "/apple-touch-icon-152x152.png",
            },
        ],
        [
            "link",
            {
                rel: "apple-touch-icon",
                sizes: "167x167",
                href: "/apple-touch-icon-167x167.png",
            },
        ],
        ["link", { rel: "manifest", href: "/site.webmanifest" }],
        ["meta", { name: "theme-color", content: "#ffffff" }],
        [
            "meta",
            { name: "msapplication-config", content: "/browserconfig.xml" },
        ],
    ],
    title: "Ink-Explorer-Well",
    description: "Notes of Exploration",
    locales: {
        root: {
            label: "简体中文",
            lang: "zh-Hans",
        },
    },
    themeConfig: {
        nav: [
            { text: "Home", link: "/" },
            { text: "Blog", link: "https://ivoink.qzz.io", target: "_blank" },
        ],
        sidebar: [
            {
                text: "Linux",
                collapsed: true,
                base: "/linux-system/",
                items: [
                    { text: "Overview", link: "/" },
                    {
                        text: "01 - Linux 安装与远程登录",
                        link: "/01-linux-installation-and-remote-login",
                    },
                    {
                        text: "02 - 软件与软件源",
                        link: "/02-software-and-software-sources",
                    },
                    {
                        text: "03 - 用户类型和用户权限",
                        link: "/03-user-types-and-user-permissions",
                    },
                    {
                        text: "04 - 文件编辑与操作",
                        link: "/04-file-editing-and-operation",
                    },
                    {
                        text: "05 - 系统与服务",
                        link: "/05-system-and-services",
                    },
                    {
                        text: "06 - Linux系统配置与安全",
                        link: "/06-configuration-and-security",
                    },
                    {
                        text: "07 - 网络配置与管理",
                        link: "/07-network-configuration-and-management",
                    },
                    {
                        text: "08 - 进程管理与日志排查",
                        link: "/08-process-management-and-log-troubleshooting",
                    },
                    {
                        text: "09 - 磁盘与存储管理",
                        link: "/09-disk-and-storage-management",
                    },
                ],
            },

            {
                text: "搭建自己的Linux开发环境",
                collapsed: true,
                base: "/build-linux-development-enviroment/",
                items: [{ text: "Overview", link: "/" }],
            },
        ],
        search: {
            provider: "local",
            options: {
                translations: {
                    button: {
                        buttonText: "搜索文档",
                        buttonAriaLabel: "搜索文档",
                    },
                    modal: {
                        displayDetails: "显示详情",
                        resetButtonTitle: "清除查询条件",
                        backButtonTitle: "返回",
                        noResultsText: "无法找到相关结果",
                        footer: {
                            selectText: "选择",
                            selectKeyAriaLabel: "输入",
                            navigateText: "切换",
                            navigateUpKeyAriaLabel: "上",
                            navigateDownKeyAriaLabel: "下",
                            closeText: "关闭",
                            closeKeyAriaLabel: "esc",
                        },
                    },
                },
            },
        },
        socialLinks: [
            { icon: "github", link: "https://github.com/ivoink/docs" },
        ],
        outline: {
            label: "页面导航",
        },
        // lastUpdated: {
        //     text: "最后更新于",
        // },
        // editLink: {
        //     text: "在 GitHub 上编辑此页面",
        // },
        docFooter: {
            prev: "上一页",
            next: "下一页",
        },

        // Translation
        editLink: false,
        lastUpdated: false,
        ignoreDeadLinks: true,
        darkModeSwitchLabel: "外观",
        lightModeSwitchTitle: "切换到浅色模式",
        darkModeSwitchTitle: "切换到深色模式",
        sidebarMenuLabel: "菜单",
        returnToTopLabel: "回到顶部",
    },
});
